fideslib/models/client.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import annotations
+
+import json
+from datetime import datetime
+from typing import Any
+
+from sqlalchemy import ARRAY, Column, ForeignKey, String
+from sqlalchemy.ext.declarative import declared_attr
+from sqlalchemy.orm import Session
+
+from fideslib.core.config import config
+from fideslib.cryptography.cryptographic_util import (
+    generate_salt,
+    generate_secure_random_string,
+    hash_with_salt,
+)
+from fideslib.cryptography.schemas.jwt import (
+    JWE_ISSUED_AT,
+    JWE_PAYLOAD_CLIENT_ID,
+    JWE_PAYLOAD_SCOPES,
+)
+from fideslib.db.base_class import Base
+from fideslib.models.fides_user import FidesUser
+from fideslib.oauth.api.scope_registry import SCOPE_REGISTRY
+from fideslib.oauth.jwt import generate_jwe
+
+DEFAULT_SCOPES: list[str] = []
+
+
+class ClientDetail(Base):  # type: ignore
+    """The persisted details about a client in the system"""
+
+    @declared_attr
+    def __tablename__(cls) -> str:  # pylint: disable=E0213
+        return "client"
+
+    hashed_secret = Column(String, nullable=False)
+    salt = Column(String, nullable=False)
+    scopes = Column(ARRAY(String), nullable=False, default="{}")
+    fides_key = Column(String, index=True, unique=True, nullable=True)
+    user_id = Column(
+        String, ForeignKey(FidesUser.id_field_path), nullable=True, unique=True
+    )
+
+    @classmethod
+    def create_client_and_secret(
+        cls,
+        db: Session,
+        scopes: list[str] | None = None,
+        fides_key: str = None,
+        user_id: str = None,
```

Comment:
Should we add some model-level tests for these methods to protect the behavior from regressing?

Reply:
I definitely want more tests. There were some places I was having a hard time with the tests right now because I need to start getting the database setup and pulling in the FastAPI parts. As I started working on that I was going beyond the first part of the ticket and pulling in a lot more things.

My end goal is to have 100% coverage once I get the other part of the ticket done.

Reply:
The plan is still to add more tests once everything is in place, but I was able to come up with some tests here without needing everything completed.